File: packages/core/src/clients/NativeClient.ts
import type { ToolDefinition } from "@langchain/core/language_models/base";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";
import { always } from "alwaysly";
import { AccessibilityTreeDiff } from "../server/accessibility/AccessibilityTreeDiff.js";
import { ChangesAnalyzerAgent } from "../server/agents/ChangesAnalyzerAgent.js";
import { RetrieverAgent } from "../server/agents/RetrieverAgent.js";
import type { UsageStats } from "../server/serverSchema.js";
import { Session } from "../server/session/Session.js";
import type { SessionId } from "../server/session/SessionId.js";
import { SessionManager } from "../server/session/SessionManager.js";
import { convertToolsToSchemas } from "../tools/toolToSchemaConverter.js";
import { getLogger } from "../utils/logger.js";
import { Client } from "./Client.js";
import { Data, looselyTypecast } from "./typecasting.js";

const logger = getLogger(import.meta.url);

export namespace NativeClient {
  export interface Props extends Client.Props {
    llm?: BaseChatModel | undefined;
  }
}

export class NativeClient extends Client {
  private sessionManager: SessionManager;
  // private model: Model;
  // private tools: Record<string, ToolClass>;
  private sessionId: SessionId;
  session: Session;

  constructor(
    props: NativeClient.Props,
    // model: Model,
    // platform: string,
    // tools: Record<string, ToolClass>,
    // llm?: BaseChatModel,
    // planner: boolean = true,
    // excludeAttributes: Set<string> = new Set(),
  ) {
    const { llm, ...superProps } = props;
    super(superProps);

    this.sessionManager = new SessionManager();

    // Convert tools to schemas for API
    const toolSchemas = convertToolsToSchemas(this.tools);
    this.sessionId = this.sessionManager.createSession({
      provider: this.model.provider,
      name: this.model.name,
      tools: toolSchemas as ToolDefinition[],
      platform: this.platform as SessionManager.CreateSessionProps["platform"],
      llm,
      planner: this.planner,
      excludeAttributes: this.excludeAttributes,
    });

    const session = this.sessionManager.getSession(this.sessionId);
    always(session);
    this.session = session;
  }

  async quit(): Promise<void> {
    this.sessionManager.deleteSession(this.sessionId);
  }

  /**
   * Plan actions to achieve a goal.
   *
   * @returns Object with explanation and steps.
   */
  async planActions(
    goal: string,
    accessibilityTree: string,
  ): Promise<Client.PlanActionsResult> {
    if (!this.session.planner) {
      return { explanation: goal, steps: [goal] };
    }

    const tree = this.session.processTree(accessibilityTree);
    const [explanation, steps] = await this.session.plannerAgent.invoke(
      goal,
      tree.toXml(this.session.excludeAttributes),
    );
    return { explanation, steps };
  }

  async addExample(goal: string, actions: string[]): Promise<void> {
    logger.debug(
      `Adding example. Goal: ${goal}, Actions: ${JSON.stringify(actions)}`,
    );
    await this.session.plannerAgent.addExample(goal, actions);
  }

  async clearExamples(): Promise<void> {
    this.session.plannerAgent.promptWithExamples.examples = [];
  }

  async executeAction(
    goal: string,
    step: string,
    accessibilityTree: string,
  ): Promise<Client.ExecuteActionResult> {
    const tree = this.session.processTree(accessibilityTree);
    const [explanation, actions] = await this.session.actorAgent.invoke(
      goal,
      step,
      tree.toXml(this.session.excludeAttributes),
    );
    return {
      explanation,
      actions: tree.mapToolCallsToRawId(actions),
    };
  }

  async retrieve(
    statement: string,
    accessibilityTree: string,
    title: string,
    url: string,
    screenshot?: string,
  ): Promise<[string, Data]> {
    const tree = this.session.processTree(accessibilityTree);
    const excludeAttrs = new Set([
      ...RetrieverAgent.EXCLUDE_ATTRIBUTES,
      ...this.session.excludeAttributes,
    ]);
    const [explanation, result] = await this.session.retrieverAgent.invoke(
      statement,
      tree.toXml(excludeAttrs),
      title,
      url,
      screenshot || null,
    );
    return [explanation, looselyTypecast(result)];
  }

  async findArea(
    description: string,
    accessibilityTree: string,
  ): Promise<Client.FindAreaResult> {
    const tree = this.session.processTree(accessibilityTree);
    const area = await this.session.areaAgent.invoke(
      description,
      tree.toXml(this.session.excludeAttributes),
    );
    return { id: tree.getRawId(area.id), explanation: area.explanation };
  }

  async findElement(
    description: string,
    accessibilityTree: string,
  ): Promise<Client.FindElementResult | undefined> {
    const tree = this.session.processTree(accessibilityTree);
    const element = (
      await this.session.locatorAgent.invoke(
        description,
        tree.toXml(this.session.excludeAttributes),
      )
    )[0];
    always(element);
    element.id = tree.getRawId(element.id);
    return element;
  }

  async analyzeChanges(
    beforeAccessibilityTree: string,
    beforeUrl: string,
    afterAccessibilityTree: string,
    afterUrl: string,
  ): Promise<string> {
    const beforeTree = this.session.processTree(beforeAccessibilityTree);
    const afterTree = this.session.processTree(afterAccessibilityTree);
    const excludeAttrs = new Set([
      ...ChangesAnalyzerAgent.EXCLUDE_ATTRIBUTES,
      ...this.session.excludeAttributes,
    ]);
    const diff = new AccessibilityTreeDiff(
      beforeTree.toXml(excludeAttrs),
      afterTree.toXml(excludeAttrs),
    );

    let analysis = "";
    if (beforeUrl && afterUrl) {
      if (beforeUrl !== afterUrl) {
        analysis = `URL changed to ${afterUrl}. `;
      } else {
        analysis = "URL did not change. ";
      }
    }

    analysis += await this.session.changesAnalyzerAgent.invoke(diff.compute());
    return analysis;
  }

  async saveCache(): Promise<void> {
    await this.session.cache.save();
  }

  async discardCache(): Promise<void> {
    await this.session.cache.discard();
  }

  async getStats(): Promise<UsageStats> {
    return this.session.stats;
  }
}

File: packages/python/src/alumnium/clients/native_client.py
from langchain_core.language_models import BaseChatModel

from ..server.accessibility import AccessibilityTreeDiff
from ..server.logutils import get_logger
from ..server.models import Model
from ..server.session_manager import SessionManager
from ..tools.base_tool import BaseTool
from ..tools.tool_to_schema_converter import convert_tools_to_schemas
from .typecasting import Data, loosely_typecast

logger = get_logger(__name__)


class NativeClient:
    def __init__(
        self,
        model: Model,
        platform: str,
        tools: dict[str, type[BaseTool]],
        llm: BaseChatModel | None = None,
        planner: bool = True,
        exclude_attributes: set[str] | None = None,
    ):
        self.session_manager = SessionManager()
        self.model = model
        self.tools = tools

        # Convert tools to schemas for API
        tool_schemas = convert_tools_to_schemas(tools)
        self.session_id = self.session_manager.create_session(
            provider=self.model.provider.value,
            name=self.model.name,
            tool_schemas=tool_schemas,
            platform=platform,
            llm=llm,
            planner=planner,
            exclude_attributes=exclude_attributes or set(),
        )

        self.session = self.session_manager.get_session(self.session_id)
        self.cache = self.session.cache

    def quit(self):
        self.session_manager.delete_session(self.session_id)

    def plan_actions(self, goal: str, accessibility_tree: str, app: str = "unknown") -> tuple[str, list[str]]:
        """
        Plan actions to achieve a goal.
        Returns:
            A tuple of (explanation, steps).
        """
        if not self.session.planner:
            return (goal, [goal])

        self.cache.app = app
        accessibility_tree = self.session.process_tree(accessibility_tree)
        return self.session.planner_agent.invoke(
            goal, accessibility_tree.to_xml(exclude_attrs=self.session.exclude_attributes)
        )

    def add_example(self, goal: str, actions: list[str]):
        logger.debug(f"Adding example. Goal: {goal}, Actions: {actions}")
        return self.session.planner_agent.add_example(goal, actions)

    def clear_examples(self):
        self.session.planner_agent.prompt_with_examples.examples.clear()

    def execute_action(
        self, goal: str, step: str, accessibility_tree: str, app: str = "unknown"
    ) -> tuple[str, list[dict]]:
        self.cache.app = app
        accessibility_tree = self.session.process_tree(accessibility_tree)
        explanation, actions = self.session.actor_agent.invoke(
            goal, step, accessibility_tree.to_xml(exclude_attrs=self.session.exclude_attributes)
        )
        return explanation, accessibility_tree.map_tool_calls_to_raw_id(actions)

    def retrieve(
        self,
        statement: str,
        accessibility_tree: str,
        title: str,
        url: str,
        screenshot: str | None,
        app: str = "unknown",
    ) -> tuple[str, Data]:
        self.cache.app = app
        accessibility_tree = self.session.process_tree(accessibility_tree)
        exclude_attrs = self.session.retriever_agent.EXCLUDE_ATTRIBUTES | self.session.exclude_attributes
        explanation, result = self.session.retriever_agent.invoke(
            statement,
            accessibility_tree.to_xml(exclude_attrs=exclude_attrs),
            title=title,
            url=url,
            screenshot=screenshot,
        )
        return explanation, loosely_typecast(result)

    def find_area(self, description: str, accessibility_tree: str, app: str = "unknown"):
        self.cache.app = app
        accessibility_tree = self.session.process_tree(accessibility_tree)
        area = self.session.area_agent.invoke(
            description, accessibility_tree.to_xml(exclude_attrs=self.session.exclude_attributes)
        )
        return {"id": accessibility_tree.get_raw_id(area["id"]), "explanation": area["explanation"]}

    def find_element(self, description: str, accessibility_tree: str, app: str = "unknown") -> dict:
        self.cache.app = app
        accessibility_tree = self.session.process_tree(accessibility_tree)
        element = self.session.locator_agent.invoke(
            description, accessibility_tree.to_xml(exclude_attrs=self.session.exclude_attributes)
        )[0]
        element["id"] = accessibility_tree.get_raw_id(element["id"])
        return element

    def analyze_changes(
        self,
        before_accessibility_tree: str,
        before_url: str,
        after_accessibility_tree: str,
        after_url: str,
        app: str = "unknown",
    ) -> str:
        self.cache.app = app
        before_tree = self.session.process_tree(before_accessibility_tree)
        after_tree = self.session.process_tree(after_accessibility_tree)
        exclude_attrs = self.session.changes_analyzer_agent.EXCLUDE_ATTRIBUTES | self.session.exclude_attributes
        diff = AccessibilityTreeDiff(
            before_tree.to_xml(exclude_attrs=exclude_attrs),
            after_tree.to_xml(exclude_attrs=exclude_attrs),
        )

        analysis = ""
        if before_url and after_url:
            if before_url != after_url:
                analysis = f"URL changed to {after_url}. "
            else:
                analysis = "URL did not change. "

        analysis += self.session.changes_analyzer_agent.invoke(diff.compute())

        return analysis

    def save_cache(self):
        self.session.cache.save()

    def discard_cache(self):
        self.session.cache.discard()

    @property
    def stats(self):
        return self.session.stats
